### PR TITLE
Hb 04

### PR DESCRIPTION
**Pull request by vremere-a**

### CODE REVIEW

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        LocalDateTime startDay = date.atStartOfDay();
+        LocalDateTime endDay = date.atTime(LocalTime.MAX);
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+            CriteriaQuery<MovieSession> criteriaQuery =
+                    criteriaBuilder.createQuery(MovieSession.class);
+            Root<MovieSession> movieSessionRoot = criteriaQuery.from(MovieSession.class);
+            Predicate movie = criteriaBuilder.equal(movieSessionRoot.get("movie3"), movieId);
+            Predicate x =
```
Comment (KostiantynVorobiov) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: I think this variable need to rename, maybe "time"

**Review by KostiantynVorobiov — APPROVED** [review]
Perfect!

File: src/main/resources/hibernate.cfg.xml [hunk 2]
```diff
@@ -12,7 +12,10 @@
         <property name="hibernate.connection.password">Vremere_a</property>
         <property name="show_sql">true</property>
         <property name="hibernate.hbm2ddl.auto">create-drop</property>
+        <property name="hibernate.enable_lazy_load_no_trans">true</property>
```
Comment (RomanDubovskyi) [inline src/main/resources/hibernate.cfg.xml]: https://vladmihalcea.com/the-hibernate-enable_lazy_load_no_trans-anti-pattern/

File: src/main/java/com/dev/cinema/model/CinemaHall.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+public class CinemaHall {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private int capacity;
+    @OneToMany
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/model/CinemaHall.java]: not sure we will need this annotation in future, let's remove that for now unless there is a particular reason to keep it

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(movieSession);
+            transaction.commit();
+            return movieSession;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't add MovieSession entity", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: let's make message more informative and add some info about entity we were trying to save same with other methods, where applicable

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        LocalDateTime startDay = date.atStartOfDay();
+        LocalDateTime endDay = date.atTime(LocalTime.MAX);
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+            CriteriaQuery<MovieSession> criteriaQuery =
+                    criteriaBuilder.createQuery(MovieSession.class);
+            Root<MovieSession> movieSessionRoot = criteriaQuery.from(MovieSession.class);
+            Predicate movie = criteriaBuilder.equal(movieSessionRoot.get("movie"), movieId);
+            Predicate x =
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: x != informative name

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import javax.persistence.criteria.Root;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+
+@Dao
+public class MovieSessionDaoImpl implements MovieSessionDao {
+    @Override
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        LocalDateTime startDay = date.atStartOfDay();
+        LocalDateTime endDay = date.atTime(LocalTime.MAX);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: we don't these variables, just pass value right away to method

File: src/main/java/com/dev/cinema/model/MovieSession.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+public class MovieSession {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/model/MovieSession.java:21]: this is for your info, but keep in mind that toOne relations have fetchType eager by default, which is not always quite good

File: src/main/java/com/dev/cinema/model/CinemaHall.java (hunk 3 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/model/CinemaHall.java]: ok

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 4 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: done

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 5 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: ok

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 6 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: ok

File: src/main/resources/hibernate.cfg.xml (hunk 2 above)
Comment (vremere-a) [reply src/main/resources/hibernate.cfg.xml]: del

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 1 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: 100&%

File: src/main/java/com/dev/cinema/service/impl/MovieSessionServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.List;
+
+@Service
+public class MovieSessionServiceImpl implements MovieSessionService {
+    @Inject
+    private MovieSessionDao movieSessionDao;
+
+    @Override
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        return movieSessionDao.findAvailableSessions(movieId,date);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/MovieSessionServiceImpl.java:18]:
```suggestion
        return movieSessionDao.findAvailableSessions(movieId, date);
```

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**